### ISSUE TITLE
packer-builder-openstack picks ipv6 over ipv4 for ssh access

### ISSUE THREAD

**Issue by rbrutas** [mitchellh/packer#3047]
when running `packer build  -var image_name_suffix=12a3d67 -var agile_network_id=xxx-xxx-xxxx-xxxx-c2e4496a4b56 packer.json`

it creates an openstack instance with both ipv4 (xx.xx.164.41) and ipv6 (xxxx:xxx:xxxx:xxx:xxxx:3eff:feec:eb64) address but packer uses the ipv6 for ssh access. i am running `packer 0.8.6`

```
2016/01/09 17:38:16 packer-builder-openstack: 2016/01/09 17:38:16 [INFO] Waiting for SSH, up to timeout: 5m0s
2016/01/09 17:38:16 ui: ==> **_cloud_openstack: Waiting for SSH to become available...
2016/01/09 17:38:16 packer-builder-openstack: 2016/01/09 17:38:16 [DEBUG] Detected address: [xxxx:xxx:xxxx:xxx:xxxx:3eff:feec:eb64]
2016/01/09 17:38:16 packer-builder-openstack: 2016/01/09 17:38:16 [DEBUG] Using IP address [xxxx:xxx:xxxx:xxx:xxxx:3eff:feec:eb64] for SSH
2016/01/09 17:38:16 packer-builder-openstack: 2016/01/09 17:38:16 [DEBUG] TCP connection to SSH ip/port failed: dial tcp [xxxx:xxx:xxxx:xxx:xxxx:3eff:feec:eb64]:22: connect: no route to host
2016/01/09 17:38:21 packer-builder-openstack: 2016/01/09 17:38:21 [DEBUG] Detected address: [xxxx:xxx:xxxx:xxx:xxxx:3eff:feec:eb64]
2016/01/09 17:38:21 packer-builder-openstack: 2016/01/09 17:38:21 [DEBUG] Using IP address [xxxx:xxx:xxxx:xxx:xxxx:3eff:feec:eb64] for SSH
2016/01/09 17:38:21 packer-builder-openstack: 2016/01/09 17:38:21 [DEBUG] TCP connection to SSH ip/port failed: dial tcp [xxxx:xxx:xxxx:xxx:xxxx:3eff:feec:eb64]:22: connect: no route to host
```

I can verify that the ipv4's ssh access is available.

**Comment by CpuID** [comment]
Why wouldnt you prefer IPv6 over IPv4 if both are available...?

I guess maybe it should be configurable in selected scenarios?

**Comment by rbrutas** [comment]
i don't think our vpn provides ipv6 address yet.

**Comment by CpuID** [comment]
Sounds like in this scenario having it configurable is probably wise. Maybe defaulting to IPv6 with an override flag to prefer IPv4 as required. As time goes on IPv6 will become more common and leaving it as the default feels right... deferring judgement to core maintainers though on this one.

**Comment by cbednarski** [comment]
Thanks for the report. We should definitely let you configure this. I think this would be at the communicator level so it will probably help everywhere.

**Comment by jcmcken** [comment]
I think I'm running into this issue with the VMware builder as well. When I try to build, I get the following error:

```
Error starting VM: VMware error: socket(): Address family not supported by protocol
```

...which seems to just be a generic ``socket`` syscall error ``EPROTONOSUPPORT``.

This is only when running ``centos-64`` (doesn't occur on ``centos``).

**Comment by rickard-von-essen** [comment]
@jcmcken Can you open a separate issue for VMware since this have been solved in v0.9.0 for OpenStack.

**Comment by rickard-von-essen** [comment]
Fixed in PR #3197

**Comment by jcmcken** [comment]
Just to close the loop for [this comment](mitchellh/packer/issues/3047#issuecomment-189516114):

I discovered my issue and what the solution was, and it's unrelated to IPv6 at all. The socket error was a red herring and was not causing ``vmrun`` to fail. 

I discovered the real issue in the ``vmware.log`` generated in the directory of the VMX file launched by ``vmrun`` (after launching ``packer build -debug`` and cancelling it prior to the directory cleanup). The error in that log looked like:

```
2016-02-29T10:05:25.383-05:00| vmx| I125: Msg_Post: Error
2016-02-29T10:05:25.383-05:00| vmx| I125: [msg.cpuid.noLongmode2] This virtual machine is configured for 64-bit guest operating systems. However, 64-bit operation is not possible.
2016-02-29T10:05:25.383-05:00| vmx| I125+ This host does not support Intel VT-x.
```

The error was caused by running ``packer`` on a VM that didn't support hardware virtualization of the CPU. Once I enabled HVM, the 64-bit build was successful.
